[WIP] Pythranized version of _solve_toeplitz.pyx

serge-sans-paille opened the pull request.

serge-sans-paille commented:
@rgommers I don't understand the remaining failure :-/

ilayn commented:
re-run the test, now passing

ilayn commented on scipy/linalg/_solve_toeplitz.py:
```diff
@@ -0,0 +1,117 @@
+        m2 = (m + 1) >> 1
+        c1 = g[m]
+        c2 = h[m]
+        for j in range(m2):
```
These are also inplace ops as far as I can see

```python
g[:m2] -= c1*h[:m2]
g[m2:] -= c2*h[:m2][::-1]  # or whatever the logic here is

ilayn commented on scipy/linalg/_solve_toeplitz.py:
```diff
@@ -0,0 +1,117 @@
+from numpy import zeros
+
+# FIXME: from numpy.linalg import LinAlgError when it's supported by Pythran
+LinAlgError = RuntimeError
```
Can it not handle subclassing of Errors?

serge-sans-paille replied:
no it cannot.

serge-sans-paille commented on scipy/linalg/_solve_toeplitz.py:
```diff
@@ -0,0 +1,117 @@
+            g_den += a[n+j-m-1] * h[m-j-1]
+
+        if g_den == 0.0:
+            raise LinAlgError("Singular principal minor")
```
That's the original error message, I don't know what I could do here.

ilayn replied:
Then nevermind, we'll revisit them at some point.

serge-sans-paille commented:
@ilayn I did most of the requested changes, And tested the performance impact, not finding any significant changes wrt. cython implementation.

@rgommers I also added support for using blas with Pythran in `meson.build`, but that was my first interaction with it, so feel free to correct me.

serge-sans-paille commented:
I've added the test case to pythran codebase to avoid regressing on it, see https://github.com/serge-sans-paille/pythran/pull/2117

serge-sans-paille commented:
The failing check seems unrelated

serge-sans-paille commented:
@ilayn any though on that new revision?

serge-sans-paille commented:
@rgommers for that PR, I've removed the call to `@` and added a FIXME. It should not be a big deal to have pythran support 1D dot product without linking to blas, but that's on the pythran side!

rgommers commented:
Thanks, that sounds good!

serge-sans-paille commented:
I don't understand the failure, otherwise looks good to me?

ilayn commented:
Nevermind that failure, it is not related but #18654 This looks good to me

serge-sans-paille commented:
Gentle ping :-)

mdhaber commented:
Hi @serge-sans-paille I wanted to help you get this merged, if you're still interested. I've taken a look at the diff between the Cython and Pythran and I see that the Pythran is mostly just a simplified version of the Cython - great! However, were there other motivations for this conversion? Considering the low maintainance needs of the function, I wasn't sure whether the simplification in the code offsets the performance regression I'm seeing for the problems that take substantial time:

On my Windows machine:

![image](https://github.com/user-attachments/assets/97474ea5-cbfa-4388-802b-b849413786b7)

And my Mac:

<img width="916" alt="image" src="https://github.com/user-attachments/assets/afd09dd7-22eb-46fa-85d0-846eded56203">

The branch I'm working with is https://github.com/mdhaber/scipy/tree/gh18628 (so I can include both versions of the function in the same build), and the script I'm using for the comparison is https://github.com/mdhaber/scipy/blob/gh18628/scipy/linalg/scratch_8286.py.

mdhaber commented:
@serge-sans-paille I thoughts I'd ping you again here. Were you still interested in merging this one? It looks like it's close to ready, but I wasn't sure it would be an improvement for users. Do you get better performance with Pythran here, or was the main benefit the simplified implementation?

serge-sans-paille commented:
@mdhaber sorry this went below my radar. I'll rebase, test, profile and then answer (the basic answer being, yes I'm interested but I need to gather digits)

serge-sans-paille commented:
I ran some benchmarks and we're faster if we use blas, slower if we don't. I'll prepare an update to the patch.

serge-sans-paille commented:
ok, so far this needs the top-of-tree pythran version to compile, and it hasn't been tagged yet :-)

mdhaber commented:
OK, please ping when the required version of Pythran is released. Is there an expected timeframe?

rgommers commented:
Re BLAS usage and this:

https://github.com/serge-sans-paille/pythran/blob/7364613cb2a955211e3974e39f04e81e35dbb71f/pythran/pythonic/numpy/dot.hpp#L15-L16

That isn't going to work I'm afraid. The name mangling is specific to the ILP64 build of `scipy-openblas`, and won't work with other BLAS libraries or LP64 OpenBLAS. To get this to work, it's going to need scipy-specific/internal code; using BLAS from C/C++ code is a bit of a pain. It should probably look like the `npy_cblas.h` and `BLAS_FUNC` usage like this:
https://github.com/scipy/scipy/blob/149f9fd32a1bdf1010944706b840c06816ea6a2f/scipy/optimize/_trlib/trlib_private.h#L42

That's all subject to change, and I'm not sure that you'd want to put that much SciPy-internals knowledge into Pythran.

serge-sans-paille commented:
@rgommers hell... how is the cython version doing then? by the way do we have minimal guarantees like SSE2 on the target platform?

rgommers commented:
> @rgommers hell... how is the cython version doing then? 

I'm afraid it gets to cheat: we have public Cython APIs (`cython_blas.pxd` and `cython_lapack.pxd`) in `scipy.linalg`. Those have very complex machinery to hide the symbol mangling, so only SciPy has to worry about that and downstream packages like scikit-learn do not. 

That Cython API uses Cython import mechanism via an extension module though, and you can't get at it from C/C++ code in SciPy:
```
% dyld_info -exports ../../build-install/lib/python3.12/site-packages/scipy/linalg/cython_blas.cpython-312-darwin.so
../../build-install/lib/python3.12/site-packages/scipy/linalg/cython_blas.cpython-312-darwin.so [arm64]:
    -exports:
        offset      symbol
        0x00002758  _PyInit_cython_blas
```
This is a pain for other C/C++ code as well. We treat that a bit too ad-hoc still.

> by the way do we have minimal guarantees like SSE2 on the target platform?

Yes, you can assume SSE3 is available, that's the baseline for both NumPy and SciPy at the moment. We do apparently have a bug there, at least there's a recent report of an AVX instruction having crept in: gh-21155.
